task_management/app/models/task.rb

```diff
@@ -9,6 +9,29 @@ class Task < ApplicationRecord
   validates :name, presence: true, length: { maximum: 50 }
   validates :deadline, presence: true
 
-  enum priority: { low: 0, normal: 1, high: 2 }
-  enum status: { todo: 0, in_progress: 1, done: 2 }
+  enum priority: { low: 1, normal: 2, high: 3 }
+  enum status: { todo: 1, in_progress: 2, done: 3 }
+
+  def self.get_sort_key(sort_item, order)
+    # ソートキーを設定
+    if sort_item.blank? || order.blank?
+      'creation_date' + ' ' + 'DESC'
+    else
+      sort_item + ' ' + order
+    end
+  end
+
+  def self.search_tasks(user_id, sort_key, condition)
+    # 検索ボタンを押下した場合
+    if condition[:search_btn] == I18n.t('tasks.button.type.search')
+      search_word = condition[:search_word]
+      status = Task.statuses.values if condition[:status] == 'all'
+      Task.where(user_id: user_id)
+          .where(status: status)
+          .where('name like ?', '%' + search_word + '%')
+          .order(sort_key)
+    else
+      Task.where(user_id: user_id).order(sort_key)
+    end
+  end
```

Comment:
ああー、すみません、指摘の仕方が悪くてかなり誤解させちゃいました汗
私としては以下のような感じでscopeを使用して書いてほしかった感じですー
```suggestion
  scope :sort_key, ->(sort_item, order) { order("? ?", sort_item, order) if (sort_item.present? || order.present?) }
  scope :hogehoge, ->(status) { where(status: status) if statuses.values.include?(status) }
  scope :search_word, ->(search_word) { where('name like ?', "%#{search_word}%") if search_word.present? }
```
テキトーに書いただけなんで動くか分からないんですけど、こんな感じ↑でスコープを定義すると以下のようにcontrollerにかけると思いますー！
```ruby
  # 既存の書き方
  def index
    user_id = User.select(:id).find_by(login_id: TEST_USER_ID)
    @order = params[:order]
    sort_key = Task.get_sort_key(params[:sort], @order)
    @tasks = Task.search_tasks(user_id, sort_key, params)
  end
```
↑が↓にのように書ける
```ruby
  # 多分こんな感じにかけると思う、たぶん
  def index
    user = User.find_by(login_id: TEST_USER_ID)
    @order = params[:order]
    @tasks = Task.where(user_id: user.id).hogehoge(status).search_word(search_word).sort_key(sort_item, @order)
  end
```
スコープの定義で後置ifなどを使用すると、その条件に当てはまらない場合はSQLで絞り込みなどが行われずにそのまま全件返される（つまりメソッドが何もしない）ため、controller（やmodel）でいろいろ条件分岐させていたものが不要になりますー

Reply:
ご説明ありがとうございます！勘違いしていました。

複雑な条件分岐をなくしてSQLでの絞り込みをする・しないができたのですね
ソースがスッキリしました！